internal/controller/apiextensions/composite/reconciler.go

```diff
@@ -595,6 +596,12 @@ func (r *Reconciler) Reconcile(ctx context.Context, req reconcile.Request) (reco
 		}
 		err = errors.Wrap(err, errCompose)
 		r.record.Event(xr, event.Warning(reasonCompose, err))
+		if kerrors.IsInvalid(err) {
+			// invalid errors could be unstable due to pointers in the error
+			// message, so we just log them, but avoid setting them as
+			// condition, to avoid endlessly looping.
+			err = errors.Wrap(errors.New(errInvalidResources), errCompose)
+		}
```

Comment:
I don't understand this bit. Could you elaborate? What would cause us to endlessly loop?

Reply:
In this specific case the Invalid error returned by the API server contains the full fmt representation of the invalid part of the resource (`ownerRefs`) which is unstable due to it containing pointers to bools that are different at every reconciliation. This means we update the XR condition with a new error every time and therefore trigger a new reconciliation loop every time.

https://github.com/crossplane/crossplane/pull/4918 would have prevented that too.

Reply:
Got it, thanks.

It's obvious to me in retrospect but explicitly mentioning that it's the API server returning unstable error strings might help.